testrepository/commands/failing.py

```diff
@@ -53,11 +55,25 @@ def _show_subunit(self, run):
         self.ui.output_stream(stream)
         return 0
 
-    def _make_result(self, repo):
+    def _make_result(self, repo, tests):
         testcommand = self.command_factory(self.ui, repo)
-        if self.ui.options.list:
+        with testcommand:
+            profiles = testcommand.profiles
+        if self.ui.options.list or self.ui.options.json:
+            def capture(test_dict):
+                test_id = test_dict['id']
+                if test_id not in tests:
+                    meta = {'profiles': []}
+                    tests[test_id] = meta
+                else:
+                    meta = tests[test_id]
+                test_profiles = profiles.intersection(test_dict['tags'])
+                meta['profiles'] = sorted(
+                    test_profiles.union(meta['profiles']))
+            analyzer = testtools.StreamToDict(on_test=capture)
             list_result = testtools.StreamSummary()
-            return list_result, list_result
+            result = testtools.CopyStreamResult([analyzer, list_result])
+            return result, list_result
         else:
             return self.ui.make_result(repo.latest_id, testcommand)
```

Comment:
This method is getting kind of complicated. Any chance you could split it up, or at least add some comments?